### PR TITLE
Refactor statespace.

### PR DESCRIPTION
**Pull request by sergeisakov**
Introduce additional base class (can be also used for unitary space). Add state size to State. Add state size checks. Better const correctness. A lot of cosmetic changes.

### CODE REVIEW

File: lib/statespace.h [hunk 1]
```diff
@@ -51,67 +45,32 @@ class StateSpace {
...
-    for_.Run(raw_size_, f, src, dest);
+  uint64_t Size() const {
+    return uint64_t{1} << num_qubits_;
   }
 
   double Norm(const State& state) const {
     auto partial_norms = static_cast<const Impl&>(*this).PartialNorms(state);
 
-    double norm = partial_norms[0];
+    double norm = partial_norms.size() > 0 ? partial_norms[0] : std::nan("");
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply lib/statespace.h]: Done.

File: lib/statespace_avx.h [hunk 2]
```diff
@@ -58,36 +58,30 @@ inline double HorizontalSumAVX(__m256 s) {
...
+  }
 
-      for (uint64_t i = 0; i < 7; ++i) {
-        s[2 * i + 1] = im[i];
-        s[2 * i + 2] = re[i];
-      }
-    };
+  bool InternalToNormalOrder(State& state) const {
+    if (state.size() != Base::raw_size_) {
+      return false;
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply lib/statespace_avx.h:80]: Yes, that is a good idea. One reason why I do not use `throw` is that I got used to follow the Google style guide and the guide basically prohibits using exceptions (qsim roughly follows the Google style guide). Anyway, this is probably beyond this PR. This mechanism of reporting errors is used everywhere in the code and it should be changed everywhere for consistency.

File: tests/statespace_testfixture.h [hunk 3]
```diff
@@ -600,23 +603,24 @@ void TestOrdering() {
...
     uint64_t size = uint64_t{1} << num_qubits;
 
-    std::vector<fp_type> vec(2 * std::max(uint64_t{8}, size), 0);
-
     StateSpace state_space(num_qubits, 1);
-    State state = state_space.CreateState(vec.data());
+    uint64_t raw_size =
+        state_space.MinimumRawSize(2 * (uint64_t{1} << num_qubits));
+    std::vector<fp_type> vec(raw_size, 0);
+    State state = state_space.CreateState(vec.data(), vec.size());
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply tests/statespace_testfixture.h:610]: Done. Merged that with the new "InvalidStateSize" test.

File: tests/statespace_testfixture.h [hunk 4]
```diff
@@ -371,7 +371,7 @@ void TestAdd() {
   state_space.SetAmpl(state2, 2, 5, 6);
   state_space.SetAmpl(state2, 3, 7, 8);
 
-  state_space.AddState(state1, state2);
+  EXPECT_TRUE(state_space.AddState(state1, state2));
```
[the remark this answers is not in the document]
Comment (sergeisakov) [reply tests/statespace_testfixture.h:374]: Done.

File: lib/statespace_avx.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (95-martin-orion) [reply lib/statespace_avx.h:80]: Sounds good - I might have confused this with the qFlex error-reporting strategy.

File: tests/statespace_testfixture.h [hunk 5]
```diff
@@ -793,6 +793,61 @@ void TestMeasurementLarge() {
...
+      state_space.MinimumRawSize(2 * (uint64_t{1} << num_qubits)) - 1;
+  std::vector<fp_type> vec(raw_size, 0);
+  State state = state_space.CreateState(vec.data(), vec.size());
+
+  EXPECT_FALSE(state_space.SetAllZeros(state));
+  EXPECT_FALSE(state_space.SetStateUniform(state));
+  EXPECT_FALSE(state_space.SetStateZero(state));
+  EXPECT_FALSE(state_space.Multiply(1.0, state));
+  EXPECT_FALSE(!std::isnan(state_space.Norm(state)));
+  EXPECT_FALSE(!state_space.PartialNorms(state).empty());
```
Comment (95-martin-orion) [inline tests/statespace_testfixture.h:816]: `EXPECT_FALSE(!<expression>)` can be replaced with `EXPECT_TRUE(<expression>)`.

Comment (sergeisakov) [reply tests/statespace_testfixture.h:816]: I just wanted to be consistent with other lines expecting false in the case of an error.

**Pull request merged**